### DISCUSSION TITLE
Question about Multilingual Support in Godot

### DISCUSSION THREAD

**Discussion by imurashka** [godotengine/godot-proposals#7936] [Scripting]
Dear Godot community,

I would like to raise an important question regarding multilingual support in Godot. While it provides developers more choices of programming languages, I am concerned that it may have a negative impact on the long-term development.

Here are some key aspects that come to mind when I think about supporting multiple languages:

1. Community Division: Supporting multiple languages can make collaboration and knowledge sharing more complex among developers using different programming languages. This may lead to division and uncertainty within the community and make it more challenging for newcomers who have to choose between different languages and tutorials.

2. Developer Mobility Limitation: Developers who have worked with Godot for several years using one language may face difficulties when transitioning to projects or teams using a different language. This can limit the mobility and career growth of developers.

3. Challenges with Packages and Libraries: When developers use different languages, it complicates the creation and sharing of packages and libraries that could be made available to everyone. Interoperability between different languages can be challenging and requires additional effort.

4. Fragmentation and Support: Supporting multiple languages also means that developers and the Godot team must invest resources in supporting each of them, which can slow down development and complicate maintenance.

Wouldn't it be more efficient to select a single programming language and collectively focus on it? This approach could potentially streamline development, foster a more unified community, and mitigate the challenges associated with multilingual support.

I think we need to discuss this situation and present clear arguments for or against my proposal. I don't believe I'm the only one pondering this question. Therefore, it would be good to discuss it and come to a single answer.

Sincerely,
Ivan

**Comment by AntonAzvazhinsky** [comment]
Thank Ivan for rising this questions. I also think we need discuss this

**Comment by gmi19941998** [comment]
I agree with Ivan; it's better to focus on one language as it will allow for a more standardized game engine. Unity also, at one point, abandoned several languages in favor of one.

**Comment by Calinou** [comment]
Here's a thread about why GDScript will remain maintained alongside C#: https://twitter.com/reduzio/status/1702582789823226206

**Comment by imurashka** [comment]
@Calinou My concern is not specifically about GDScript itself, but rather the potential presence of many other languages alongside it. If it's not feasible to choose an existing popular language instead of GDScript for the engine, would it be worth considering sticking with GDScript as the primary and only language?

**Comment by timothyqiu** [comment]
If I understand it correctly, the goal is to move C# support into GDExtension so that we eventually have only two options: GDScript and GDExtension as a general interface for other languages.

Since GDScript and GDExtension share the same API, they can call each other seamlessly, and I would call it a single language :P

Languages made available via GDExtension are maintained by the community I think, so it would not slow down core development nor complicate maintenance.

**Comment by Enderlook** [comment]
Wouldn't be even better to **only** expose the `GDExtension` interface?
Then C#, C++, Rust, GDScript, etc could be added as an addon by connecting with this interface.
This would provide several benefits:
1) Unity codebase: no longer requires special code just for GDScript or C#, instead, there is a single interface to be maintained by the core engines.
2) Equality: if all languages share the same interface then all of them can do the same functionality. Godot would no longer have a preferred child.
3) Free improvements for all: this also means that if for example, the GDExtensions are not *good enough* for GDScript, then in order to improve GDScript, developers are "forced" to improve GDExtensions first, meaning that all other languages can benefit from this improvements free. In other words, the upgrades of a language can easily be shared by others.
4) Faster iteration: if languages are separated from the actual engine, language support doesn't need to iterate at the same speed as the engine. As far as the GDExtensions don't have any breaking changes, addons for other languages are not bound to a specific Godot version, which allows faster upgrades. Basically, the teams which develop support for languages are not bound by the team which develops the engine itself.
